src/hotspot/share/runtime/handshake.hpp

```diff
@@ -123,6 +153,31 @@ class HandshakeState {
   ProcessResult try_process(HandshakeOperation* match_op);
 
   Thread* active_handshaker() const { return _active_handshaker; }
+
+  // Suspend/resume support
+ private:
+  // This flag is true when the thread owning this
+  // HandshakeState (the _handshakee) is suspended.
+  volatile bool _suspended;
+  // This flag is true while there is async handshake (trap)
+  // on queue. Since we do only need one, we can reuse it if
+  // thread gets suspended again (after a resume)
+  // and we have not yet processed it.
+  bool _async_suspend_handshake;
+
+  // Called from the suspend handshake.
+  bool suspend_with_handshake();
+  // Called from the async handshake (the trap)
+  // to stop a thread from continuing execution when suspended.
+  void do_self_suspend();
+
+  bool is_suspended()                       { return Atomic::load(&_suspended); }
+  void set_suspended(bool to)               { return Atomic::store(&_suspended, to); }
+  bool has_async_suspend_handshake()        { return _async_suspend_handshake; }
+  void set_async_suspend_handshake(bool to) { _async_suspend_handshake = to; }
```

Comment:
Does this need to be an Atomic::store?

Reply:
No, _async_suspend_handshake is now only loaded/stored with mutex held.